implement formula guessing, C-finite case

rwst opened the pull request:
This allows the following:

```
    >>> from sympy.solvers.guesser import guess
    >>> guess([1,2,3,4,5,6])
    n + 1
    >>> guess([0,1,1,2,3,5,8,13])
    sqrt(5)*(1/2 + sqrt(5)/2)**n/5 - sqrt(5)*(-sqrt(5)/2 + 1/2)**n/5
    >>> guess([2,3,5,9,17,33,65])
    2**n + 1
    >>> guess([1/3,1/9,1/27,1/81,1/243,1/729])
    0.333333333333333*0.333333333333333**n

    >>> guess([1*x,3*x,6*x,10*x,15*x,21*x,28*x])
    x*(n**2/2 + 3*n/2 + 1)
    >>> guess([1,y+x,y**2+2*x*y+x**2,y**3+3*x*y**2+3*x**2*y+x**3,(y+x)**4,(y+x)**5])
    (x + y)**n
    >>> guess([x,y,-x+2*y,-2*x+3*y,-3*x+4*y,-4*x+5*y])
    n*(-x + y) + x
    >>> guess([x,y,x+y,x+2*y,2*x+3*y,3*x+5*y])
    sqrt(5)*(1/2 + sqrt(5)/2)**n*(-x + sqrt(5)*x + 2*y)/10 + (-sqrt(5)/2 + 1/2)**n*(sqrt(5)*x/10 + x/2 - sqrt(5)*y/5)
```

rwst committed:
implement recurrence finding, sequence guessing, C-finite case

skirpichev commented on sympy/solvers/guesser.py:
```diff
@@ -0,0 +1,81 @@
+    Journal 31.1-2 (2013): 23-32.
+"""
+
+def guess(seq):
```
better name?

skirpichev commented on sympy/solvers/guesser.py:
```diff
@@ -0,0 +1,81 @@
+
+def guess(seq):
+    """
+    Return closed form depending on `n` that produces the sequence,
```
Why not remove out trivial steps, involving rsolve and return instead a recurrence?

skirpichev commented on sympy/solvers/guesser.py:
```diff
@@ -0,0 +1,81 @@
+.. [1] Zeilberger, Doron. "The C-finite ansatz." The Ramanujan
+    Journal 31.1-2 (2013): 23-32.
+"""
+
```
newline

skirpichev commented on sympy/solvers/tests/test_guesser.py:
```diff
@@ -0,0 +1,15 @@
+from sympy import Symbol, sqrt
+from sympy.solvers.guesser import guess
+from sympy.abc import x,y
+
```
newline

skirpichev commented on sympy/solvers/tests/test_guesser.py:
```diff
@@ -0,0 +1,15 @@
+
+def test_guesser():
+    n = Symbol('n', integer=True)
+    assert guess([1,2,3,4,5,6]) == n + 1
```
space after comma

skirpichev commented on sympy/solvers/tests/test_guesser.py:
```diff
@@ -0,0 +1,15 @@
+    assert guess([1*x,3*x,6*x,10*x,15*x,21*x,28*x]) == x*(n**2/2 + 3*n/2 + 1)
+    assert guess([1,y+x,y**2+2*x*y+x**2,y**3+3*x*y**2+3*x**2*y+x**3,
+                  (y+x)**4,(y+x)**5]) == (x + y)**n
+    assert guess([x,y,-x+2*y,-2*x+3*y,-3*x+4*y,-4*x+5*y]) == n*(-x + y) + x
```
no end of line

skirpichev commented on sympy/solvers/guesser.py:
```diff
@@ -0,0 +1,81 @@
+    coeffs = trim_zeros(K.transpose().tolist()[0][::-1])
+    cl = len(coeffs)
+    if cl == l:
+        return None
```
this line isn't covered

skirpichev commented on sympy/solvers/guesser.py:
```diff
@@ -0,0 +1,81 @@
+    try:
+        K = A.nullspace()[0]
+    except IndexError:
+        return None
```
Not covered as well (btw, you could just use "return")

skirpichev commented on sympy/solvers/guesser.py:
```diff
@@ -0,0 +1,81 @@
+
+    l = len(seq)
+    if l < 6:
+        raise ValueError('Sequence too short for guessing. Please give at least 6 values.')
```
you should test this too.

skirpichev commented on sympy/solvers/guesser.py:
```diff
@@ -0,0 +1,81 @@
+    Return None if no result is found.
+
+    Does only C-finite sequences currently.
+    >>> from sympy.solvers.guesser import guess
```
add newline and please fix examples (see doctests failure).

asmeurer commented:
I would call this sequence guessing, not formula guessing. It only guesses formulas for functions over nonnegative integers.

asmeurer commented:
Don't know if it's worth mentioning in the docstring, but another good way to "guess" a formula from a sequence is to look it up in OEIS.

skirpichev commented:
> I would call this sequence guessing, not formula guessing.

In fact, it's recurrence guessing, and very specific kind off.

Not sure if it worth a separate module.  recurr.py seems to be a good place.

rwst commented:
OK, let's say we want a function returning a recurrence. But there is no recurrence object. It doesn't suffice to give an equation, because a recurrence also may have start values. So, in order to make the result usable by not only rsolve, but also allow evaluation, asymptotics, generating function etc, a core recurrence class would be best.

skirpichev commented:
> function returning a recurrence

recurrence relation

> It doesn't suffice to give an equation, because a recurrence also may have start values.

Why in this case it doesn't suffice?  It's trivial to provide initial conditions for rsolve from the first terms of the seq.

skirpichev commented:
#8267 - another case we want to find a recurrence relation.

rwst commented:
I addressed all your remarks and did extensive changes. Also now I know how to squash commits without a new PR, so I did that. I didn't manage to add the two function's docs to the documentation. At least here, they weren't found (I did `make clean; make` in `doc/`).

skirpichev commented on sympy/solvers/recurr.py:
```diff
@@ -822,3 +822,126 @@ def rsolve(f, y, init=None):
     return solution
+
+
+def rfind(seq):
```
I'm ok with something like this function.  But you should think about naming.

Also, I think it's a bad idea to use hard-coded Symbol's for the output.  You can make optional arguments for this.

rwst replied:
> But you should think about naming.

I did. As long as you don't offer an alternative I will stick to "rfind".

skirpichev commented on sympy/solvers/recurr.py:
```diff
@@ -822,3 +822,126 @@ def rsolve(f, y, init=None):
+        Journal 31.1-2 (2013): 23-32.
+
+    See Also
+    ========
```
add newline

skirpichev commented on sympy/solvers/recurr.py:
```diff
@@ -822,3 +822,126 @@ def rsolve(f, y, init=None):
+    ========
+    guess_seq
+
+    >>> from sympy.solvers.recurr import rfind
```
I think, this should go "Examples" section.  And you should move it above (before References).

skirpichev commented on sympy/solvers/recurr.py:
```diff
@@ -822,3 +822,126 @@ def rsolve(f, y, init=None):
+
+    """
+    from sympy import Symbol, Function, Eq, simplify
+    coeffs = _rfind_helper(seq)
```
this helper is useless, if you will have rfind only

skirpichev commented on sympy/solvers/recurr.py:
```diff
@@ -822,3 +822,126 @@ def rsolve(f, y, init=None):
+        cl = len(coeffs)
+        eq = sum(coeffs[i]*a(n+i) for i in range(cl))
+        start = {a(i): seq[i] for i in range(cl-1)}
+        return rsolve(eq, a(n), start)
```
No newline at the end of file.  Please [run tests](https://github.com/sympy/sympy/wiki/Development-workflow#be-sure-that-all-tests-of-sympy-pass).

rwst committed:
cosmetics addressing reviewer's comments

rwst commented:
Thanks for the review.

skirpichev commented on sympy/solvers/__init__.py:
```diff
@@ -11,7 +11,8 @@
 
-from .recurr import rsolve, rsolve_poly, rsolve_ratio, rsolve_hyper
+from .recurr import rsolve, rsolve_poly, rsolve_ratio, rsolve_hyper, \
+    rfind, guess_seq
```
-1 to this, unless you will have better names.

skirpichev commented on sympy/solvers/recurr.py:
```diff
@@ -822,3 +822,134 @@ def rsolve(f, y, init=None):
     return solution
+
+
+def rfind(seq, rname='a'):
```
Better is to use symbols.

skirpichev commented on sympy/solvers/recurr.py:
```diff
@@ -822,3 +822,134 @@ def rsolve(f, y, init=None):
+    coeffs = _rfind_helper(seq)
+    if coeffs is not None:
+        a = Function(rname)
+        n = Symbol('n', integer=True)
```
still hardcoded

skirpichev commented on sympy/solvers/recurr.py:
```diff
@@ -822,3 +822,134 @@ def rsolve(f, y, init=None):
+    return trim_zeros(K.transpose().tolist()[0])
+
+
+def guess_seq(seq):
```
-1 to this.  I think, you should include something like rfind only.

rwst replied:
I have opened a discussion at https://groups.google.com/forum/?hl=en#!topic/sympy/UCb028Rxih8 days ago.
Please tell me your reasons there.

czgdp1807 commented:
This idea is nice but is there something like this already in `master`?
